proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,1066 @@
+start subscribing with dynamic parameters constraints like the following:
+
+```textproto
+// For {env=prod, version=v1}
+{key_constraints:[
+  {key:"env" value:{
+    constraints:[{value:"prod"}]
+  }},
+  {key:"version" value:{
+    constraints:[{value:"v1"}]
+  }}
+]}
+```
+
+The server or cache has to match that set of constraints against the
+existing sets of dynamic parameters, which do not specify the `version`
+key at all.
+
+Conversely, if the new key is added on the server side first, then the
+server will have resource variants with parameters like this:
+- `{env=prod, version=v1}`
+- `{env=prod, version=v2}`
+- `{env=test, version=v1}`
+- `{env=test, version=v2}`
+
+But at this point, the clients are continuing to subscribe without
+constraints on this new key.  So the server or cache needs to figure out
+(e.g.) which of the first two sets of constraints to use for constraints
+that require `env` to be `prod` but do not specify `version`.
+
+We address this transition scenario by allowing the set of constraints
+for a given key to match any resource variant that does not specify that
+key at all.  This allows constraints for new keys to be added on clients
+before the corresponding keys are added on the resources on the server, but
+it does introduce some additional ambiguity into the matching.  For example,
+let's say that the server has the following two variants of a resource:
+- `{env=prod}`
+- `{env=prod, version=v1}`
+
+Now consider what happens if a client subscribes with the following
+constraints:
+
+```textproto
+{key_constraints:[
+  {key:"env" value:{
+    constraints:[{value:"prod"}]
+  }},
+  {key:"version" value:{
+    constraints:[{value:"v1"}]
+  }}
+]}
+```
+
+These constraints can match either of the above variants of the resource.
+This situation can be avoided by establishing a best practice that all
+variants of a given resource must have the same set of keys.
+
+There is still a possible ambiguity that can occur if a server adds
+multiple variants of a new key that clients are not yet sending.
+For example, let's say that the server has the following two variants
+of a resource:
+- `{env=prod, version=v1}`
+- `{env=prod, version=v2}`
+
+Consider what happens if a client subscribes with the following constraints:
+
+```textproto
+{key_constraints:[
+  {key:"env" value:{
+    constraints:[{value:"prod"}]
+  }}
+]}
+```
+
+These constraints can match either variant of the above resource.
+This can be avoided by establishing a best practice of not adding multiple
+variants of a new parameter until clients are sending the new parameter.
+However, if this does happen, the cache implementation is free to pick
+one of the variants at random.
+
+So, the expected order of changes for this kind of transition would be:
+1. Change clients to start sending a constraint for `version=v1`.
+2. Add the dynamic parameter `version=v1` to all existing resources.
+3. Create new variants of each resource with `version=v2`.
```

Comment:
I think so, although @fuqianggao can chime in as well.

I would expect the normal use-case here to be that the user configures their clients to start sending a constraint for the new key before they start differentiating their resources based on that key.  Doing the reverse doesn't really make sense, because there isn't yet anything to differentiate.

To say this another way, I think it could be quite common for users to divide their clients into categories based on some key, even if they initially don't configure those sets of clients differently from each other.  But the fact that they have that in place means that they can easily change the config of a subset of their clients simply by changing the configuration reesources in the control plane.

Reply:
Makes sense, I don't disagree with the theory here, I'm thinking maybe something empirical to back this from TD as an example would make me more confident.